### ISSUE TITLE
programmatic API

### ISSUE THREAD

**Issue by Ladicek** [smallrye/smallrye-fault-tolerance#370]
In addition to the CDI interceptor binding annotations specified by MP Fault Tolerance, we could also support a programmatic API. Note that it would be very different from the core API, which I very much consider an implementation detail! I'm thinking something like this:

```java
Callable<String> = FaultTolerance.create()
    .withFallback(...)
    .withRetry(...)
    .withTimeout(...)
    .withBulkhead(...)
    .withCircuitBreaker(...)
    .from(... some Callable<String> ...);

Callable<CompletionStage<String>> = FaultTolerance.createAsync()
    .withFallback(...)
    .withRetry(...)
    .withTimeout(...)
    .withBulkhead(...) // this one is different from the synchronous API!
    .withCircuitBreaker(...)
    .from(... some Callable<CompletionStage<String>> ...);
```

We would probably still honor the fault tolerance strategy nesting as specified by MP Fault Tolerance, and would probably ignore the `Future` style of async invocations.

### Open questions

Integration with MP Config?
- ignore, users would have to configure everything explicitly through the API;
- if present, use for default values (no class+method combination lookup, just the "annotation" default) that are not configured explicitly through the API (the API would always have priority);
- alternative of the previous option, but the API would provide a way to set an "identifier" that we would use to lookup the configuration (the identifier could possibly also be a class+method name combination).

Integration with MP Metrics / Micrometer (see also #265)?
- ignore;
- use if present.

How to support additional async types (`Single`, `Uni` etc.)?
- ignore (bad);
- add explicit API for each supported type (also quite bad);
- `FaultTolerance.createAsync().from(...)` would take `Object` (also quite bad);
- hmm not sure if there's any other option.

How to support blocking and non-blocking async calls?
- `FaultTolerance.createBlocking` and `FaultTolerance.createNonBlocking`;
- `FaultTolerance.createAsync(boolean)`;
- something else?

How to register programmatically created circuit breakers into `CircuitBreakerMaintenance`?
- ignore them from `CircuitBreakerMaintenance` perspective (bad, `CircuitBreakerMaintenance.resetAll` is very useful for testing);
- use `CDI.current` to find `CircuitBreakerMaintenance` if it exists (have to deal with CDI missing);
- create a separate registry which `CircuitBreakerMaintenance` could somehow access.

Programmatic access to `CircuitBreakerMaintenance`?
- no;
- yes, but completely separate from the CDI variant;
- somehow integrated (?).

**Comment by Ladicek** [comment]
I've been pondering this for a while, and I think the biggest question to resolve is: should this be integrated with the CDI-based API? I think there are 3 possible answers:

- never, it's completely separate;
- always, it's an error if there's no CDI;
- when possible.

The last option is clearly the most desirable one, but also the most complex one. It has itself 2 variants:

- eager, when CDI presence is determined upfront and not revised later;
- lazy, when CDI integration is performed on demand.

Again, the last option is the most desirable one, but also the most complex one.

What would require CDI integration?

- MP Metrics
  - without CDI, we wouldn't emit metrics at all
  - the situation might be different with Micrometer?
- `CircuitBreakerMaintenance`
  - we would probably want something like `FaultTolerance.circuitBreakerMaintenance()`, which would return an object that would delegate to the CDI bean if possible, but would still work in absence of CDI

Notably, MP Config wouldn't require CDI presence, because it is itself designed to be used before CDI is available. So we could easily use that API (`ConfigProvider.getConfig()` etc.).

**Comment by Ladicek** [comment]
The initial version just merged in #552 has 2 different implementations: completely standalone and CDI integrated.

Here are answers to the open questions from the 1st post above.

> Integration with MP Config?

None in the standalone impl. None yet in the CDI impl, with the exception of `MP_Fault_Tolerance_NonFallback_Enabled`. May change in the future.

> Integration with MP Metrics / Micrometer (see also #265)?

None at the moment. Will likely change in the future in the CDI impl.

> How to support additional async types (`Single`, `Uni` etc.)?

None yet. Will likely change in the future. I can see a separate API called `MutinyFaultTolerance`, structured similarly to the async part of `FaultTolerance`.

> How to support blocking and non-blocking async calls?

There's an option to enable thread offload for async calls (`withThreadOffload(boolean)`).

> How to register programmatically created circuit breakers into CircuitBreakerMaintenance?

The CDI impl looks up the `CircuitBreakerMaintenance` from CDI and uses that, so it's completely integrated. No integration in the standalone impl.

> Programmatic access to CircuitBreakerMaintenance?

Yes, there's `FaultTolerance.circuitBreakerMaintenance()`. In the CDI impl, it's integrated.
